### PR TITLE
Document and create new pandas example

### PR DESCRIPTION
**Pull request by franchuterivera**

### CODE REVIEW

File: doc/manual.rst [hunk 1]
```diff
@@ -92,6 +93,21 @@ Resampling strategies
...
+
+If your data contains categorical values (in the features or targets), autosklearn will automatically encode your data using a `sklearn.preprocessing.LabelEncoder <https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html>`_ for unidimensional data and a `sklearn.preprocessing.OrdinalEncoder <https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html>`_ for multidimensional data. Target vectors (y_train/y_test) are automatically encoded for classification if using one of the aforementioned supported formats (test and train classes are encoded in a common space to prevent unseen classes during testing). In the case of features (X_train, X_test) your data must be provided as a pd.DataFrame for the automatic encoding to trigger. Additionally, the categorical columns of such pandas DataFrame must have boolean or category dtype (`Working with categorical data <https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html>`_). Numerical columns will preserve their dtype. Other pandas DataFrame dtypes (Object, time series) must be first converted to a valid categorical or numerical dtype.
```
Comment (mfeurer) [inline doc/manual.rst:107]: Could you please make this two paragraphs, one for the features and one for the targets? Also, could you please note that numpy arrays with a list of data types are accepted, too?

Comment (franchuterivera) [reply doc/manual.rst:107]: Understood, that will make it clearer.

File: doc/manual.rst [hunk 2]
```diff
@@ -92,6 +93,21 @@ Resampling strategies
...
+Supported Inputs
+================
+*auto-sklearn* can accept targets for the following tasks (more details on `Sklearn algorithms <https://scikit-learn.org/stable/modules/multiclass.html>`_):
+* Binary Classification
+* Multiclass Classification
+* Multilabel Classification
+* Regression
+* Multioutput Regression
+
+You can provide both a features/targets training pair (X_train/y_train) and an optional testing pair (X_test/Y_test) as exemplified in `Pandas Train and Test inputs <examples/example_pandas_train_test.html>`_. Supported formats for these training and testing pairs are: np.ndarray, pd.DataFrame, scipy.sparse.csr_matrix and python lists.
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply doc/manual.rst:105]: I added a clarification on the need for testing data in both the example and the manual page. Please let me know if this brief description is enough!

File: examples/example_pandas_train_test.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+# ===================================
+
+predictions = cls.predict(X_test)
+print("Accuracy score", sklearn.metrics.accuracy_score(y_test, predictions))
+
+############################################################################
+# Plot the ensemble performance
+# ===================================
+
+ensemble_performance_frame = pd.DataFrame(cls._automl[0].ensemble_performance_history)
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply examples/example_pandas_train_test.py:68]: I added this plot!

File: examples/example_pandas_train_test.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+############################################################################
+# Data Loading
+# ============
+
+# Using Australian dataset https://www.openml.org/d/40981
+X, y = sklearn.datasets.fetch_openml(data_id=40981, return_X_y=True, as_frame=True)
+
+# bool and category will be automatically encoded.
+# Targets for classification are also automatically encoded
+# If using fetch_openml, data is already properly encoded, below
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply examples/example_pandas_train_test.py:28]: Done!

File: examples/example_pandas_train_test.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# -*- encoding: utf-8 -*-
+"""
+===============================
+Test and Train data with Pandas
+===============================
+
+*auto-sklearn* can automatically encode categorical columns using a label/ordinal encoder. This example highlights how to properly set the dtype in a DataFrame for this to happen, and showcase how to input also testing data to autosklearn.
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply examples/example_pandas_train_test.py:7]: Done!

File: examples/example_get_pipeline_components.py [hunk 6]
```diff
@@ -34,6 +34,10 @@
     resampling_strategy='holdout',
     # We want autosklearn to use pca as preprocessor
     include_preprocessors=['pca'],
+    # Bellow two flags are provided to speed up calculations
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply examples/example_get_pipeline_components.py:37]: Done!

File: requirements.txt [hunk 7]
```diff
@@ -3,6 +3,7 @@ Cython
 
 numpy>=1.9.0
 scipy>=0.14.1
+matplotlib
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply requirements.txt:6]: Done!

File: examples/example_feature_types.py [hunk 8]
```diff
@@ -38,8 +37,11 @@
 # ==========================
 
 cls = autosklearn.classification.AutoSklearnClassifier(
-    time_left_for_this_task=120,
-    per_run_time_limit=30,
+    time_left_for_this_task=60,
```
Comment (mfeurer) [inline examples/example_feature_types.py:40]: Looking at this examples, it could actually be drastically simplified - we shouldn't need the label encoder any more, and it should be sufficient to download the dataset from OpenML only once. Also, could you please add a link to the pandas example from here and vice versa as they are very related?

File: examples/example_pandas_train_test.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import autosklearn.classification
+from autosklearn.metrics import accuracy
+
+
+def get_runhistory_models_performance(automl):
+    metric = cls._automl[0]._metric
+    data = automl._automl[0].runhistory_.data
+    performance_list = []
+    for run_key, run_value in data.items():
+        if 'train_loss' not in run_value.additional_info:
```
Comment (mfeurer) [inline examples/example_pandas_train_test.py:37]: Would it be possible to check for SMAC's `StatusType` here?

File: examples/example_pandas_train_test.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            # Ignore crashed runs
+            continue
+        endtime = pd.Timestamp(time.strftime('%Y-%m-%d %H:%M:%S',
+                                             time.localtime(run_value.endtime)))
+        train_score = metric._optimum - (metric._sign * run_value.additional_info['train_loss'])
+        test_score = metric._optimum - (metric._sign * run_value.additional_info['test_loss'])
+        performance_list.append({
+            'Timestamp': endtime,
+            'indv_model_train_score': train_score,
```
Comment (mfeurer) [inline examples/example_pandas_train_test.py:46]: Sorry, I guess I said training loss, but actually we'd be interested in the validation loss as that's what we're optimizing.

Comment (franchuterivera) [reply examples/example_pandas_train_test.py:46]: oh, I see, I will use the cost directly then

File: examples/example_pandas_train_test.py (hunk 9 above)
Comment (franchuterivera) [reply examples/example_pandas_train_test.py:37]: Yes, I will query for success then (all other cases like abort, crash and so on are not useful )

File: examples/example_feature_types.py (hunk 8 above)
Comment (franchuterivera) [reply examples/example_feature_types.py:40]: Done!

**Comment by franchuterivera** [comment]
I had to move away from Adult to Australian as adult had NaN values in categorical columns.

File: autosklearn/data/validation.py [hunk 11]
```diff
@@ -406,10 +406,19 @@ def _check_and_encode_target(
 
             # Mypy redefinition
             assert self.target_encoder is not None
-            self.target_encoder.fit(y)
+
+            # remove ravel warning from pandas Series
+            if len(y.shape) > 1 and y.shape[1] == 1 and hasattr(y, "values"):
+                self.target_encoder.fit(y.values.ravel())
```
Comment (mfeurer) [inline autosklearn/data/validation.py:412]: Could you please use 'to_numpy' here instead of 'values'? By that it is clear what the return value is, see [here](https://pandas.pydata.org/pandas-docs/stable/whatsnew/v0.24.0.html#accessing-the-values-in-a-series-or-index).

File: autosklearn/data/validation.py [hunk 12]
```diff
@@ -406,10 +406,19 @@ def _check_and_encode_target(
 
             # Mypy redefinition
             assert self.target_encoder is not None
-            self.target_encoder.fit(y)
+
+            # remove ravel warning from pandas Series
+            if len(y.shape) > 1 and y.shape[1] == 1 and hasattr(y, "values"):
```
Comment (mfeurer) [inline autosklearn/data/validation.py:411]: Is there a reason why this is only an issue for pandas series and not numpy arrays?

File: examples/example_pandas_train_test.py (hunk 10 above)
Comment (mfeurer) [reply examples/example_pandas_train_test.py:53]: Could you please also rename this to validation score?

File: examples/example_feature_types.py [hunk 13]
```diff
@@ -7,31 +7,35 @@
...
-data = sklearn.datasets.fetch_openml(data_id=179, as_frame=True)
-feat_type = ['Categorical' if x.name == 'category' else 'Numerical' for x in data['data'].dtypes]
+# Auto-sklearn can automatically recognize categorical/numerical data from a pandas
+# DataFrame. This example highlights how the user can provide the feature types,
+# which is particularly important when using numpy arrays, as there is no
+# per-column dtype in this case. Auto-sklearn will honor the argument feat_type
+# in any case.
+# feat_type is a list that tags each column from a DataFrame/ numpy array / list
+# with the case-insensitive string categorical or numerical, accordingly.
+feat_type = ['Categorical' if x.name == 'category' else 'Numerical' for x in X.dtypes]
```
Comment (mfeurer) [inline examples/example_feature_types.py:38]: Wait, does this mean we can pass a dataframe and `feat_type` at the same time? I think we should prohibit this and allow only one of those. What I think would be great if you could download the dataset as a numpy array as before, but as a so-called bunch object (i.e. without `return_X_y`) and extract the information from there.

File: doc/manual.rst [hunk 14]
```diff
@@ -102,11 +102,16 @@ Supported Inputs
...
+* You can provide a pandas DataFrame, with properly formatted columns. If a column has numerical dtype, *auto-sklearn* will not encode it and it will be passed directly to scikit-learn. If the column has a categorical/boolean class, it will be encoded. If the column is of any other type (Object or Timeseries), an error will be raised. For further details on how to properly encode your data, you can check the example `Working with categorical data <https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html>`_). If you are working with time series, it is recommended that you follow this approach `Working with time data <https://stats.stackexchange.com/questions/311494/>`_.
+
+Regarding the targets (y_train/y_test), if the task involves a classification problem, such features will be automatically encoded. It is recommended to provide both y_train and y_test during fit, so that a common encoding is created between these splits (if only y_train is provided during fit, the categorical encoder will not be able to handle new classes that are exclusive to y_test). If the task is regression, no encoding happens on the features.
```
Comment (mfeurer) [inline doc/manual.rst:114]: For regression I assume it should be 'targets' and not 'features', right?

**Comment by mfeurer** [comment]
Hm, also there is now an issue in one of the examples which appears to be related to the changes: https://travis-ci.org/github/automl/auto-sklearn/jobs/715099392

File: examples/example_pandas_train_test.py (hunk 10 above)
Comment (franchuterivera) [reply examples/example_pandas_train_test.py:53]: I think we should also rename the ensemble to be validation.

But I found a legacy code that we need to discuss. Y_valid in the ensemble builder doesn't make sense, as it relies on datamanager to have x_valid. This is no longer the case right?

I think this should be a separate PR, to prune up this Y_valid.

For this PR I will just rename the column. Let me know if this makes sense

File: doc/manual.rst (hunk 14 above)
Comment (franchuterivera) [reply doc/manual.rst:114]: sorry, I will make the change

File: autosklearn/data/validation.py (hunk 11 above)
Comment (franchuterivera) [reply autosklearn/data/validation.py:412]: I see, thanks I will move to this

File: autosklearn/data/validation.py (hunk 12 above)
Comment (franchuterivera) [reply autosklearn/data/validation.py:411]: Yeah, for numpy:
https://github.com/automl/auto-sklearn/blob/28305d414ee3e0656cdf8ded8f949865cb50699d/autosklearn/data/validation.py#L137

For pandas, if we ravel the values so early we lose the column/dtype info.

File: examples/example_pandas_train_test.py (hunk 10 above)
Comment (mfeurer) [reply examples/example_pandas_train_test.py:53]: > I think we should also rename the ensemble to be validation.

Yes and no. There it's actually the training performance and not the validation performance. 

> But I found a legacy code that we need to discuss. Y_valid in the ensemble builder doesn't make sense, as it relies on datamanager to have x_valid. This is no longer the case right?

Yes, I think you can remove all the logic here.

> For this PR I will just rename the column. Let me know if this makes sense

Okay. How about calling the strings 'single_best_optimization_score', 'single_best_test_score', 'ensemble_optimization_score' and 'ensemble_test_score'?

File: examples/example_feature_types.py (hunk 13 above)
Comment (franchuterivera) [reply examples/example_feature_types.py:38]: Understood, I will add a value error if both feat_types and pandas object is passed. Also update the example accordingly.

**Comment by franchuterivera** [comment]
I cannot reproduce the example failure:
```
python examples/example_feature_types.py
Accuracy score 0.8786127167630058
```
I will push the changes and see if this is reproducible. Then, act accordingly

File: examples/example_pandas_train_test.py (hunk 10 above)
Comment (franchuterivera) [reply examples/example_pandas_train_test.py]: thanks for the suggestion, I will go with this naming

File: autosklearn/automl.py [hunk 15]
```diff
@@ -347,7 +347,13 @@ def fit(
                                      'valid feature types, you passed `%s`' % ft)
 
         # Feature types dynamically understood from dataframe
-        if feat_type is None and self.InputValidator.feature_types:
+        if feat_type is not None and self.InputValidator.feature_types:
```
Comment (mfeurer) [inline autosklearn/automl.py:350]: Could you please also add a unit test for this?

File: autosklearn/data/validation.py (hunk 12 above)
Comment (mfeurer) [reply autosklearn/data/validation.py]: Okay, got it. What does actually happen if the validator was trained with a numpy array and we then pass a pandas array or the other way round? Should we have unit tests for that?

File: examples/example_feature_types.py [hunk 16]
```diff
@@ -7,39 +7,48 @@
...
 # Data Loading
 # ============
-# Load adult dataset from openml.org, see https://www.openml.org/t/2117
-X, y = sklearn.datasets.fetch_openml(data_id=179, return_X_y=True)
-
-# y needs to be encoded, as fetch openml doesn't download a float
-y = preprocessing.LabelEncoder().fit_transform(y)
+# Load Australian dataset from https://www.openml.org/d/40981
+bunch = data = sklearn.datasets.fetch_openml(data_id=40981, as_frame=True)
+y = y = bunch['target'].to_numpy()
```
Comment (mfeurer) [inline examples/example_feature_types.py:29]: One `y =` should be enough here.

File: autosklearn/automl.py (hunk 15 above)
Comment (franchuterivera) [reply autosklearn/automl.py:350]: Yes thanks, I will add unit test/support to make sure that the dtype does not change once fitted.

File: examples/example_feature_types.py (hunk 16 above)
Comment (franchuterivera) [reply examples/example_feature_types.py]: sorry, fixed
